src/d3a/models/market/blockchain_interface.py

```diff
@@ -59,11 +72,81 @@ def bc_listener(self):
         pass
 
 
+class SubstrateBlockchainInterface():
+    def __init__(self):
+        self.contracts = {'main': main_address}
+        self.substrate = SubstrateInterface(
+            url=default_url,
+            address_type=42,
+            type_registry_preset='default'
+        )
+
+    def compose_call(self, data, contract_address):
+        call = self.substrate.compose_call(
+            call_module='Contracts',
+            call_function='call',
+            call_params={
+                'dest': contract_address,
+                'value': 1 * 10**18,
+                'gas_limit': 1000000000000,
+                'data': data
+            }
+        )
+        return call
+
+    def simple_contract_call(self, contract_address, function, path):
+        data = get_function_metadata(function, path)
+        call = self.compose_call(data, contract_address)
+        return call
+
+    def storage_contract_call(self, contract_address, function, path):
+        trade_id = random.randint(10, 32)
+        logger.info('TRADE ID {}'.format(trade_id))
+        data = get_function_metadata(function, path) + hex2(trade_id) + \
+            address_to_hex(ALICE_ADDRESS) + address_to_hex(BOB_ADDRESS) + \
+            swap_byte_order(hex2(test_value)) + '000000000000000000' + hex2(test_rate)
+        call = self.compose_call(data, contract_address)
+        return trade_id, call
+
+    def create_new_offer(self, energy, price, seller):
+        return str(uuid.uuid4())
+
+    def cancel_offer(self, offer):
+        pass
+
+    def change_offer(self, offer, original_offer, residual_offer):
+        pass
+
+    def handle_blockchain_trade_event(self, offer, buyer, original_offer, residual_offer):
+        return str(uuid.uuid4()), residual_offer
+
+    def track_trade_event(self, trade):
+        path = Path.cwd().joinpath('src', 'd3a', 'models', 'market', 'metadata.json')
+        trade_id, call = self.storage_contract_call(
+            self.contracts['main'],
+            "trade",
+            str(path)
+        )
+        keypair = Keypair.create_from_mnemonic(mnemonic)
+        extrinsic = self.substrate.create_signed_extrinsic(call=call, keypair=keypair)
+
+        try:
+            result = self.substrate.submit_extrinsic(extrinsic, wait_for_inclusion=True)
+            print("Extrinsic '{}' sent and included in block '{}'".format(
```

Comment:
Please do not forget to replace prints with logging.debug before merging, thanks!

Reply:
Ah yes, this is a code block from the interface, will fix